pkg/arvo/app/chat-push-hook.hoon

```diff
@@ -0,0 +1,339 @@
+            =(our.bowl (provider-for src.bowl rid))
+        ==
+      ~&  [%redirect src.bowl %to (provider-for src.bowl rid)]
+      (redirect rid (provider-for src.bowl rid))
+    ::  Provide directly
+    =/  box  (chat-scry:store bowl path)
+    ?~  box  !!
+    [%give %fact ~ %chat-update !>([%create path])]~
+  --
+::
+++  on-agent
+  |=  [=wire =sign:agent:gall]
+  ^-  (quip card _this)
+  |^
+  ?+  -.sign                  (on-agent:def wire sign)
+      %kick  ?+  wire         !!
+          [%permissions ~]    kick-permissions
+          [%store @ *]        (kick-store wire)
+      ==
+      %watch-ack              (watch-ack wire p.sign)
+      %fact  ?+  p.cage.sign  (on-agent:def wire sign)
+          %chat-update        (fact-chat-update !<(update:store q.cage.sign))
+          %permission-update  (fact-perm-update !<(permission-update q.cage.sign))
+  ==  ==
+  ::
+  ++  kick-permissions
+    :_  this
+    [%pass /permissions %agent [our.bowl %permission-store] %watch /updates]~
+  ::
+  ++  kick-store
+    |=  wire=(lest @ta)
+    ~&  store-kick+wire
+    ?.  (~(has by providers) (path-to-rid t.wire))  [~ this]
+    ~&  %chat-store-resubscribe
+    =/  mailbox=(unit mailbox:store)
+      (chat-scry:store bowl t.wire)
+    :_  this
+    [%pass wire %agent [our.bowl %chat-store] %watch [%mailbox t.wire]]~
+  ::
+  ++  watch-ack
+    |=  [=^wire tang=(unit tang)]
+    ^-  (quip card _this)
+    ?~  tang  [~ this]
+    ?.  ?=([%store *] wire)  [~ this]
+    =/  pok  [%poke %chat-push-hook-action !>([%remove t.wire])]
+    :_  this
+    [%pass / %agent [our.bowl %chat-push-hook] pok]~
+  ::
+  ++  fact-chat-update
+    |=  =update:store
+    ^-  (quip card _this)
+    ?>  (team:title our.bowl src.bowl)
+    ?+  -.update     [~ this]
+        %delete
+      =/  =rid  (path-to-rid path.update)
+      ?.  (~(has by providers) rid)  [~ this]
+      =.  providers  (~(del by providers) rid)
+      :_  this
+      [%pass [%mailbox path.update] %agent [our.bowl %chat-store] %leave ~]~
+    ::
+        %message
+      :_  this
+      [%give %fact [%mailbox path.update]~ %chat-update !>(update)]~
+    ::
+        %messages
+      :_  this
+      [%give %fact [%mailbox path.update]~ %chat-update !>(update)]~
+    ==
+  ::
+  ++  fact-perm-update
+    |=  fact=permission-update
+    ^-  (quip card _this)
+    |^
+    :_  this
+    ?+  -.fact   ~
+        %add     (handle-permissions %add path.fact who.fact)
+        %remove  (handle-permissions %remove path.fact who.fact)
+    ==
+    ++  handle-permissions
+      |=  [kind=?(%add %remove) =path who=(set ship)]
+      ^-  (list card)
+      %-  zing
+      %+  turn
+        (~(app-paths-from-group metadata bowl) %chat path)
+      |=  chat=^path
+      ^-  (list card)
+      ?.  (~(has by providers) (path-to-rid chat))  ~
+      %-  zing
+      %+  turn  ~(tap in who)
+      |=  =ship
+      ?:  (is-permitted bowl ship chat)
+        ::  XX If they are permitted, under what cicumstances will it be
```

Comment:
If they are permitted, they don't ever need to be removed. But notice that the case in which we do nothing on 300 is when they were previously on a blacklist and we have removed them from that blacklist.

Reply:
So if they're being removed from a blacklist, that's a %remove fact as well? Got it.